File: scripts/publish.mjs
/**
 * Publish the built LX-DSH installer to the update server in one step.
 *
 * Usage: node scripts/publish.mjs [installer-path]
 *   - installer-path defaults to dist/LX-DSH Setup <version>.exe
 *   - credentials come from lx-dsh/.env.publish (LX_UPDATE_ADMIN_USER /
 *     LX_UPDATE_ADMIN_PASS) or the environment; that file is git-ignored
 *
 * The endpoint is the update server's one-shot publish API: it creates or
 * updates the version record (summary = RELEASE_NOTES text, notes = bullet
 * list) and upserts the win asset in one multipart call, so a release is one
 * command after `pnpm run dist:full`.
 */
import { readFileSync, existsSync, statSync } from 'node:fs'
import { join, dirname } from 'node:path'
import { fileURLToPath } from 'node:url'

const root = join(dirname(fileURLToPath(import.meta.url)), '..')

// Credentials: environment first, then .env.publish (KEY=VALUE lines).
function loadEnvPublish() {
  const path = join(root, '.env.publish')
  if (!existsSync(path)) return {}
  const out = {}
  for (const line of readFileSync(path, 'utf8').split('\n')) {
    const m = /^([A-Z0-9_]+)=(.*)$/.exec(line.trim())
    if (m !== null) out[m[1]] = m[2]
  }
  return out
}
const env = { ...loadEnvPublish(), ...process.env }
const BASE = env.LX_UPDATE_SERVER ?? 'http://123.57.129.111'
const USER = env.LX_UPDATE_ADMIN_USER
const PASS = env.LX_UPDATE_ADMIN_PASS
if (USER === undefined || PASS === undefined) {
  console.error('publish: set LX_UPDATE_ADMIN_USER / LX_UPDATE_ADMIN_PASS (env or lx-dsh/.env.publish)')
  process.exit(1)
}

const pkg = JSON.parse(readFileSync(join(root, 'package.json'), 'utf8'))
const version = pkg.version
// Only the releasing version's own section reaches the update dialog: the
// file keeps one cumulative ledger, but the changelog a user sees while
// updating from x.y.z must describe x.y.(z+1) alone, not the whole history.
const fullNotes = existsSync(join(root, 'RELEASE_NOTES.md'))
  ? readFileSync(join(root, 'RELEASE_NOTES.md'), 'utf8').trim()
  : ''
const sectionHeader = new RegExp(`^# LX-DSH ${version.replace(/\./g, '\\.')} 更新日志\\s*$`, 'm')
const sectionStart = fullNotes.search(sectionHeader)
const notesText = sectionStart === -1
  ? fullNotes
  : fullNotes
    .slice(sectionStart)
    .split(/^# LX-DSH \d+\.\d+\.\d+ 更新日志\s*$/m)
    .filter(part => part.trim() !== '')
    .at(0)
    ?.trim() ?? ''
const bullets = notesText.split('\n').filter(line => line.startsWith('- ')).map(line => line.slice(2).trim())

const installer = process.argv[2] ?? join(root, 'dist', `LX-DSH Setup ${version}.exe`)
if (!existsSync(installer)) {
  console.error(`publish: installer missing: ${installer} — build it first (pnpm run dist:full)`)
  process.exit(1)
}
const size = statSync(installer).size

// Login for the bearer token.
const loginRes = await fetch(`${BASE}/api/admin/login`, {
  method: 'POST',
  headers: { 'Content-Type': 'application/json' },
  body: JSON.stringify({ username: USER, password: PASS }),
})
if (!loginRes.ok) {
  console.error(`publish: login failed ${loginRes.status}`)
  process.exit(1)
}
const { token } = await loginRes.json()

// One-shot publish: multipart file + version metadata.
const form = new FormData()
const bytes = readFileSync(installer)
form.set('file', new Blob([bytes]), `LX-DSH-Setup-${version}.exe`)
form.set('version', version)
form.set('channel', 'stable')
form.set('date', new Date().toISOString().slice(0, 10))
form.set('summary', notesText)
form.set('notes', JSON.stringify(bullets))
form.set('platform', 'win')
const publishRes = await fetch(`${BASE}/api/admin/publish`, {
  method: 'POST',
  headers: { Authorization: `Bearer ${token}` },
  body: form,
})
const publishBody = await publishRes.json()
if (!publishRes.ok) {
  console.error(`publish: failed ${publishRes.status}`, publishBody)
  process.exit(1)
}
console.log(`published: ${publishBody.filename} (${(publishBody.size / 1048576).toFixed(1)} MB)`)

// Verify the public manifest reflects the release before declaring success.
const latest = await (await fetch(`${BASE}/update/win/latest.json`)).json()
if (latest.version !== version || latest.size !== size) {
  console.error(`publish: manifest mismatch — served v${latest.version} size ${latest.size}, expected v${version} size ${size}`)
  process.exit(1)
}
console.log(`verified: /update/win/latest.json serves v${latest.version}${latest.notes === null || latest.notes === undefined ? '' : ` with changelog (${latest.notes.length} chars)`}`)

File: update-server/publish.py
import requests, json, sys, os

# ── publish.py — publish a release to the LX-DSH update server ──────────────────
#
# Uploads both the full NSIS installer AND the lightweight update package (delta
# zip) for a given version.  The installer is served via /win/latest.yml for the
# electron-updater full-fallback path; the update package is served via
# /update/win/latest.json for the delta update path.
#
# Usage:
#   python publish.py <version> <installer.exe> <update.zip> <base_version>
#
#   version        e.g. 0.3.0
#   installer.exe  dist/LX-DSH Setup 0.3.0.exe   (full NSIS installer)
#   update.zip     dist/LX-DSH-update-0.3.0.zip  (delta update package)
#   base_version   e.g. 0.2.0  (the version the delta was built against;
#                  pass "none" if this is the first release / full snapshot)
#
# Optional env overrides (all read from update-server/.env, git-ignored):
#   LX_UPDATE_SERVER   e.g. http://<host>
#   ADMIN_USER
#   ADMIN_PASS

def _load_env():
    env_path = os.path.join(os.path.dirname(os.path.abspath(__file__)), '.env')
    if os.path.exists(env_path):
        for line in open(env_path, encoding='utf-8'):
            line = line.strip()
            if line and not line.startswith('#') and '=' in line:
                k, v = line.split('=', 1)
                os.environ.setdefault(k.strip(), v.strip())

_load_env()
SERVER = os.environ.get("LX_UPDATE_SERVER")
USER = os.environ.get("ADMIN_USER")
PASS = os.environ.get("ADMIN_PASS")
if not SERVER or not USER or not PASS:
    sys.exit("LX_UPDATE_SERVER / ADMIN_USER / ADMIN_PASS missing — set them in update-server/.env (git-ignored) or the environment")

if len(sys.argv) < 4:
    print("Usage: python publish.py <version> <installer.exe> <update.zip> [base_version]")
    sys.exit(1)

version = sys.argv[1]
installer_path = sys.argv[2]
update_path = sys.argv[3]
base_version = sys.argv[4] if len(sys.argv) > 4 and sys.argv[4] != "none" else None

# ── login ──────────────────────────────────────────────────────────────────────
r = requests.post(SERVER + "/api/admin/login", json={"username": USER, "password": PASS}, timeout=15)
token = r.json()["token"]
headers = {"Authorization": "Bearer " + token}
print("logged in")

# ── create or reuse version ────────────────────────────────────────────────────
# Check if the version already exists; if so reuse its id, otherwise create it.
r = requests.get(SERVER + "/api/versions", headers=headers, timeout=15)
existing = {v["version"]: v["id"] for v in r.json()}

if version in existing:
    vid = existing[version]
    print(f"version {version} already exists (id={vid}), reusing")
else:
    version_data = {
        "version": version,
        "channel": "stable",
        "summary": f"LX-DSH v{version}",
        "date": __import__("datetime").date.today().strftime("%Y.%m.%d"),
        "notes": []
    }
    r = requests.post(SERVER + "/api/admin/versions", json=version_data, headers=headers, timeout=15)
    vid = r.json()["id"]
    print(f"created version {version}, id: {vid}")

# ── upload full installer (kind=portable) ──────────────────────────────────────
size = os.path.getsize(installer_path)
print(f"\nuploading installer {installer_path} ({size/1048576:.1f} MB)...")
with open(installer_path, "rb") as f:
    r = requests.post(
        SERVER + f"/api/admin/versions/{vid}/upload",
        data={"platform": "win", "kind": "portable"},
        files={"file": (f"LX-DSH-{version}-win.exe", f)},
        headers=headers,
        timeout=600
    )
print("installer upload:", r.json())

# ── upload update package (kind=update, baseVersion=...) ──────────────────────
size = os.path.getsize(update_path)
print(f"\nuploading update package {update_path} ({size/1048576:.1f} MB)...")
with open(update_path, "rb") as f:
    data = {"platform": "win", "kind": "update"}
    if base_version:
        data["baseVersion"] = base_version
    r = requests.post(
        SERVER + f"/api/admin/versions/{vid}/upload",
        data=data,
        files={"file": (f"LX-DSH-update-{version}-win.zip", f)},
        headers=headers,
        timeout=600
    )
print("update upload:", r.json())

# ── verify ────────────────────────────────────────────────────────────────────
print("\n── verifying ──")

r = requests.get(SERVER + "/api/versions/latest", timeout=15)
print("latest version:", json.dumps(r.json(), ensure_ascii=False, indent=2)[:400])

r = requests.get(SERVER + "/win/latest.yml", timeout=15)
print("\n/win/latest.yml (full installer):")
print(r.text)

r = requests.get(SERVER + "/update/win/latest.json", timeout=15)
print("/update/win/latest.json (delta package):")
print(json.dumps(r.json(), ensure_ascii=False, indent=2))
